### ISSUE TITLE
It doesn't work

### ISSUE THREAD

**Issue by myself600** [spillerrec/qt-flif-plugin#2]
I've successfully compiled `libflif.so` and placed it in `/usr/lib/qt4/plugins/imageformats`, and also made sure `/usr/lib/libflif.so{,.0}` are present. However _Gwenview_ returns

>Loading ‘Room 1.flif’ failed
Gwenview cannot display documents of type application/octet-stream.

I suppose this is how it should work, right?

**Comment by spillerrec** [comment]
Be aware that plugins compiled for Qt 4 doesn't work with Qt 5 applications, and similarly the other way around. So also make sure you place the plugin in the folder which correspond to the correct Qt version.
If you used cmake to compile, it will be compiled for Qt 5. If you use QMake, it will depend on the version. Run `qmake --version`, and it will say which version of Qt it is using. To get the other version, use `qmake4`, `qmake-qt4` or whatever naming scheme your distro is using.

For Gwenview you need to register FLIF as an mime-type in xdg. You can use the XML file in /qt-flif-plugin/configuration/ for that. If you are using a new version of Gwenview, this should be all you need to do, otherwise you need to do a lot of trickery with KDE in order to recognize it.

**Comment by myself600** [comment]
>If you used cmake to compile, it will be compiled for Qt 5.

I've used QMake per the instructions.

>Run qmake --version, and it will say which version of Qt it is using.

<pre>
QMake version 2.01a
Using Qt version 4.8.6 in /usr/lib/x86_64-linux-gnu
</pre>

>To get the other version, use qmake4, qmake-qt4 or whatever naming scheme your distro is using.

`qt4-qmake` is what I used for compilation.

>For Gwenview you need to register FLIF as an mime-type in xdg. You can use the XML file in /qt-flif-plugin/configuration/ for that.

Ran `xdg-mime install configuration/x-flif.xml`. _Gwenview_ now returns

<pre>
Loading ‘Room 1.flif’ failed
Gwenview cannot display documents of type image/x-flif.
</pre>

>If you are using a new version of Gwenview, this should be all you need to do, otherwise you need to do a lot of trickery with KDE in order to recognize it.

<pre>
$ gwenview -v
Qt: 4.8.6
KDE Development Platform: 4.13.3
Gwenview: 4.13.1
</pre>

**Comment by spillerrec** [comment]
Yeah, that is a nearly 3 years old version of Gwenview... Please be aware that Qt 5 was released 5 years ago, and that Qt 4 is no longer supported by The Qt Company, so it is no longer receiving any security patches. At that note, be aware that Qt 4.8.6 has known vulnerabilities in its BMP code.

It looks like you did everything so far correctly. You need to copy the file `/configuration/qimageioplugins/x-flif.desktop` to a KDE4 directory which I believe is `/usr/share/kde4/servicetypes/`. You might need to put in the sub-folder `qimageioplugins/`, but I don't remember. (And I can't check since Gwenview no longer depends on it.) As a quick look inside the file will show, it just tells KDE that there exists a Qt image plugin that can read and write files with that mime-type.

**Comment by myself600** [comment]
>Yeah, that is a nearly 3 years old version of Gwenview...

Sure, I know but that's just what my distro comes with. Anyway, I've installed it only for viewing WebP files since _gThumb_ has a performance lag.

>At that note, be aware that Qt 4.8.6 has known vulnerabilities in its BMP code.

Hmm… I never use BMP for images since it's uncompressed, so I'm unsure whether I should bother.

>You need to copy the file /configuration/qimageioplugins/x-flif.desktop to a KDE4 directory which I believe is /usr/share/kde4/servicetypes/. You might need to put in the sub-folder qimageioplugins/, but I don't remember. (And I can't check since Gwenview no longer depends on it.)

Tried both. Gwenview still shows _No Plugin Found_ (under _Plugins_ tab).

I'll try to use a recent Qt 5/Gwenview version and then report back.

**Comment by spillerrec** [comment]
> Sure, I know but that's just what my distro comes with.

I assumed so as well, so just a heads up that that you might want to reconsider your state of affairs. I remember running Debian for a while, but having to change due to its insistence on old buggy software in their fear of introducing new bugs.
A distro which hasn't updated Qt 4 to 4.8.7 by now (released nearly 2 years ago as the last Qt 4 version) is worthless. It only backports fixes (thus cannot introduce new bugs), so it is more secure, stable, and less buggy. Furthermore it is binary compatible, so there is no way it can introduce issues.

> Hmm… I never use BMP for images since it's uncompressed, so I'm unsure whether I should bother.

Well, images you create yourself are never an issue, as you have to craft such files very specifically to cause any harm. But if you ever load an image from someone else (e.g. the internet) you could be in risk, and the file extension can easily be faked.
Again, just a heads up to show that it is not just hypothetically that there could potentially be some bug, they are there, publicly documented, and even in code Gwenview actually uses.

> Tried both. Gwenview still shows No Plugin Found (under Plugins tab).

You could try out some of the other desktop files in the configuration folder, some might be needed. If you haven't restarted (or killed all KDE processes), you might want to try that as well, and any potential other caches of desktop files.
I did get it to work somehow back then, but it really took me many hours to get it working and I'm not quite sure exactly was is required and what is not.

You can also try to set the environment variable `QT_DEBUG_PLUGINS` to non-zero when starting Gwenview (works with both Qt 4 and 5). It will spit out a lot of information when the program starts, and it will tell if Qt have found and accepted the plugin. If it is in the correct folder, it will appear in the list, and tell if it fails to load it. If it appears and it says it is working, the issue is with KDE or Gwenview not being configured properly.

**Comment by myself600** [comment]
Apologies for the delay, but some things happened which I couldn't foresee.

>You can also try to set the environment variable QT_DEBUG_PLUGINS to non-zero when starting Gwenview […] and it will tell if Qt have found and accepted the plugin.

It seems that Gwenview does recognize the plugin:
<pre>
QFactoryLoader::QFactoryLoader() looking at "/usr/lib/qt4/plugins/imageformats/libflif.so" 
keys ("flif") 
</pre>

>I'll try to use a recent Qt 5/Gwenview version and then report back.

Unfortunately, I couldn't build a Gwenview version newer than 4.14.3 (with Qt 4.8.6, I haven't bothered upgrading to 4.8.7) due to a missing dependency `kf5-kio-dev`, which will only going to be available with the next OS release. I can try to build it on a virtual machine but that would be pointless since I can't use it on my system anyway. So I'm giving up on Gwenview.

I'm currently trying other image viewers.

**Comment by spillerrec** [comment]
You could try my minimalistic image viewer: https://github.com/spillerrec/imgviewer

But I should really try to get some virtual machines running with a few common distributions which I can reroll to a known state to know exactly what to I need to do to get it working with different applications. What distro are you using btw? If it is not something obscure I might want to try to use that and see if I can get old Gwenview to work again.
